### ISSUE TITLE
Faster Building

### ISSUE THREAD

**Issue by mitkof6** [opensim-org/opensim-core#682]
It is minor issue, but I find it productive when the library builds fast. I was curious, why OpenSim library takes a lot of time to build in contrast to Simbody, which builds quite fast. I had a quick search how to improve the building time and found the following guidelines:

A discussion on [Stackoverflow](http://stackoverflow.com/questions/373142/what-techniques-can-be-used-to-speed-up-c-compilation-times).

Also there is the [Unity Builds](http://buffered.io/posts/the-magic-of-unity-builds/), but I am not sure if it is a good choice.

In summary:
1. Pimpl Idiom
2. Forward Declarations
3. Unity Build

**Comment by chrisdembia** [comment]
Thanks for the suggestions. I think the main reason OpenSim takes longer to compile than Simbody is because it is larger. I agree it's a real pain to compile OpenSim, and decreasing the build time will make the community better.

- Forward declarations is definitely something we could improve. If you see includes that can be turned into forward declarations, please submit a PR!
- I've tried unity builds (seems like a hack) and they didn't work out of the box so I gave up (they are not guaranteed to work for all projects). [cotire](https://github.com/sakra/cotire) helps here.
- I've also tried precompiled headers and the speed up was not significant (used [cotire](https://github.com/sakra/cotire)).
- PIMPL only speeds up incremental builds; it probably increases the build time of building the whole project.

Is your problem with building the whole project the first time or with incremental builds?

For now though, try using clang. Clang compiles our code up to 2 times faster than gcc or MSVC's compiler. You can also try using an LLVM frontend compiler with MSVC; I'm not sure if that will cause any speed up though.

**Comment by chrisdembia** [comment]
If using MSVC, you might be able to get faster build times by compiling the projects in parallel (instead of simply the objects within each project in parallel). I know how to do this on the command line only. From the build directory:

```
cmake --build . --target install --config RelWithDebInfo -- /maxcpucount:8
```

**Comment by chrisdembia** [comment]
I just found this google project that might help us find unnecessary #include's or #include's that could be converted to forward declarations: https://github.com/include-what-you-use/include-what-you-use.

It doesn't seem too easy to use, so I haven't tried it out yet. @mitkof6 if you get curious and you try `include-what-you-use`, let me know.

**Comment by mitkof6** [comment]
> I just found this google project that might help us find unnecessary #include's or #include's that could be converted to forward declarations: https://github.com/include-what-you-use/include-what-you-use; http://include-what-you-use.org/

Let me check it. I was thinking of a script that parses the files and checks for forward declaration.

**Comment by chrisdembia** [comment]
> For eclipse Includator

Nice find; looks easier to use than `include-what-you-use`.

**Comment by chrisdembia** [comment]
clang/gcc `-ftime-report`

**Comment by mitkof6** [comment]
I have begun using unity builds in my projects, and I am very satisfied with the results. Despite the fact that when I change a single .cpp the project will build all the cpps, still the building time is very good (2-3s for 10 files i7). Moreover, OpenSim project is partitioned into libraries so the number of cpps is not very big. I have found a nice [cmake](https://cheind.wordpress.com/2009/12/10/reducing-compilation-time-unity-builds/) function that generates unity_build.cpp files:

function(enable_unity_build UB_SUFFIX SOURCE_VARIABLE_NAME)
    set(files ${${SOURCE_VARIABLE_NAME}})
    # Generate a unique filename for the unity build translation unit
    set(unit_build_file ${CMAKE_CURRENT_BINARY_DIR}/${UB_SUFFIX}_UnityBuild.cpp)
    # Exclude all translation units from compilation
    set_source_files_properties(${files} PROPERTIES HEADER_FILE_ONLY true)
    # Open the ub file
    file(WRITE ${unit_build_file} "// Unity Build generated by CMake\n")
    # Add include statement for each translation unit
    foreach(source_file ${files} )
        file( APPEND ${unit_build_file} "#include <${CMAKE_CURRENT_SOURCE_DIR}/${source_file}>\n")
    endforeach(source_file)
    # Complement list of translation units with the name of ub
    set(${SOURCE_VARIABLE_NAME} ${${SOURCE_VARIABLE_NAME}} ${unit_build_file} PARENT_SCOPE)  
endfunction(enable_unity_build)

example use:

#################################################################################################

set(Target TaskSpace)

set(INCS 
    TaskSpaceDLL.h
    AccelerationSaturationController.h
    DynamicCompensator.h
    IController.h
    IOptimizationTarget.h
    MuscleCompensator.h
    MuscleOptimization.h
    PDController.h
    PositionTask.h
    SpatialTask.h
    SynergyOptimizationTarget.h
    TaskBasedController.h
    TaskDynamics.h
    TaskDynamicsPrioritization.h
    TaskPrimitive.h
    TaskSpaceInverseKinematics.h
    VelocitySaturationController.h)
set(SRCS 
    AccelerationSaturationController.cpp
    DynamicCompensator.cpp
    MuscleCompensator.cpp
    MuscleOptimization.cpp
    PDController.cpp
    PositionTask.cpp
    SpatialTask.cpp
    SynergyOptimizationTarget.cpp
    TaskBasedController.cpp
    TaskDynamics.cpp
    TaskDynamicsPrioritization.cpp
    TaskPrimitive.cpp
    TaskSpaceInverseKinematics.cpp
    VelocitySaturationController.cpp)

#################################################################################################

#################################################################################################

if(USE_UNITY_BUILD)
    # Comment to disable unit build
    enable_unity_build(${Target} SRCS)
endif()

add_library(${Target} SHARED ${INCS} ${SRCS})

target_link_libraries(${Target} ${OpenSim_LIBRARIES})
set_target_properties(${Target}         
        PROPERTIES        
        PROJECT_LABEL "${Target}"  
        FOLDER "Libraries"      
)

**Comment by chrisdembia** [comment]
Thanks for sharing. Have you looked into cotire?

**Comment by mitkof6** [comment]
I am testing it, I managed to build each cotire target (libraries) separately, but the all_unity target skips builds.

`1>------ Skipped Build: Project: osimCommon_unity, Configuration: Debug x64 ------
1>Project not selected to build for this solution configuration 
2>------ Skipped Build: Project: osimLepton_unity, Configuration: Debug x64 ------
2>Project not selected to build for this solution configuration 
3>------ Skipped Build: Project: osimSimulation_unity, Configuration: Debug x64 ------
3>Project not selected to build for this solution configuration 
4>------ Skipped Build: Project: osimActuators_unity, Configuration: Debug x64 ------
4>Project not selected to build for this solution configuration 
5>------ Skipped Build: Project: osimAnalyses_unity, Configuration: Debug x64 ------
5>Project not selected to build for this solution configuration 
6>------ Skipped Build: Project: osimTools_unity, Configuration: Debug x64 ------
6>Project not selected to build for this solution configuration 
========== Build: 0 succeeded, 0 failed, 2 up-to-date, 6 skipped ==========
`

[OpenSim forked](https://github.com/mitkof6/opensim-core/tree/cotire)

**Comment by jenhicks** [comment]
Closing given age, but feel free to open new concerns.
